### PR TITLE
fix: PureCN rule

### PR DESCRIPTION
**Pull request by khurrammaqbool**
## Description

PureCN rule fails due to bash strict mode when purity and ploidy can not be estimated. This causes the whole pipeline to terminate prematurely. 

#### Fixed

- PureCN command to bypass bash strict mode.

## Documentation

- [x] N/A
- [ ] Updated Balsamic documentation to reflect the changes as needed for this PR.
    - [Document Name]

### Feature Tests

- [ ] N/A
- [x] Test [Description]
    - [Screenshot]

### Pipeline Integrity Tests

- **Report deliver (generation of the `.hk` file)**
    - [x] N/A
    - [ ] Verified
- **TGA T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **TGA T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **UMI T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **UMI T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **WGS T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **WGS T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **QC Workflow**
    - [x] N/A
    - [ ] Verified
- **PON Workflow**
    - [x] N/A
    - [ ] Verified

## Clinical Genomics Stockholm

<details>
<summary></summary>

### Documentation

- **Atlas documentation**
    - [x] N/A
    - [ ] Updated: [Link]
- **Web portal for Clinical Genomics**
    - [x] N/A
    - [ ] Updated: [Link]

### User Changes

- [x] N/A
- [ ] This PR affects the output files or results.
    - [ ] User feedback is considered unnecessary because [Justification].
    - [ ] Affected users have been included in the development process and given a chance to provide feedback.

### Infrastructure Changes

- **Stored files in Housekeeper**
    - [x] N/A
    - [ ] Updated: [Link]
- **CG (CLI and delivered/uploaded files)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Servers (configuration files on Hasta)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Scout interface**
    - [x] N/A
    - [ ] Updated: [Link]

</details>

## Checklist

> [!IMPORTANT]  
> Ensure that all checkboxes below are ticked before merging.

### For Developers

- **PR Description**
    - [x] Provided a comprehensive description of the PR.
    - [ ] Linked relevant user stories or issues to the PR.
- **Documentation**
    - [ ] Verified and updated documentation if necessary.
- **Tests**
    - [x] Described and tested the functionality addressed in the PR.
    - [x] Ensured integration of the new code with existing workflows.
    - [ ] Confirmed that meaningful unit tests were added for the changes introduced.
    - [x] Checked that the PR has successfully passed all relevant code smells and coverage checks.
- **Review**
    - [x] Addressed and resolved all the feedback provided during the code review process.
    - [ ] Obtained final approval from designated reviewers.

### For Reviewers

- **Code**
    - [ ] Code implements the intended features or fixes the reported issue.
    - [ ] Code follows the project's coding standards and style guide.
- **Documentation**
    - [ ] Pipeline changes are well-documented in the CHANGELOG and relevant documentation.
- **Tests**
    - [ ] The author provided a description of their manual testing, including consideration of edge cases and boundary
      conditions where applicable, with satisfactory results.
- **Review**
    - [ ] Confirmed that the developer has addressed all the comments during the code review.

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -1,3 +1,12 @@
+[x.x.x]
+-------
+
+Fixed:
+^^^^^^
+* PureCN fail due to bash strict mode https://github.com/Clinical-Genomics/BALSAMIC/pull/
```
Comment (ivadym) [inline CHANGELOG.rst:6]:
```suggestion
* PureCN fail due to bash strict mode https://github.com/Clinical-Genomics/BALSAMIC/pull/1404
```

File: CHANGELOG.rst [hunk 2]
```diff
@@ -1,3 +1,12 @@
+[x.x.x]
```
Comment (ivadym) [inline CHANGELOG.rst:1]:
```suggestion
[14.0.1]
```

File: BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_normal_tga.rule [hunk 3]
```diff
@@ -188,6 +188,11 @@ rule purecn_call_CNV_research:
 export PURECN='/opt/PureCN/PureCN.R'
 
 # Run PureCN to estimate tumor-purity and ploidy
+
+# Keep exit_status to bypass bash stict mode and continue even if purecn fails.
+exit_status="true"
```
Comment (ivadym) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_normal_tga.rule:193]: bash supports booleans, why not using it here?

**Review by mathiasbio — COMMENTED** [review]
Looks good! Nice job! 🌟  Have you done any tests to verify that it's not crashing anymore, rerunning the case in production for instance? :)

**Review by ivadym — COMMENTED** [review]
Nice ⭐ 

Please update the base branch to `release_v14.0.1` before merging. Since we'll need to do a verification, you could test your PR directly there, and use the case we encountered issues with in production.

And don't forget to update the default comment in the table to indicate that the purity ploidy calculation has failed. Include it also in the changelog. It would be nice to notify customers about this in the email.

**Comment by khurrammaqbool** [comment]
Screenshot 2024-02-29 at 13 07 15

**Comment by khurrammaqbool** [comment]
> Looks good! Nice job! 🌟 Have you done any tests to verify that it's not crashing anymore, rerunning the case in production for instance? :)

Thank you! I tested the solution on the same case that failed in production and the screenshot shows that the run completed without crashing on the error.

File: BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_normal_tga.rule [hunk 4]
```diff
@@ -188,6 +188,11 @@ rule purecn_call_CNV_research:
...
+
+{{
```
Comment (ivadym) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_normal_tga.rule:195]:
```suggestion
```

File: BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_normal_tga.rule [hunk 5]
```diff
@@ -199,7 +204,8 @@ Rscript $PURECN \
 --genome {params.genome} \
 --fun-segmentation Hclust \
 --force --post-optimize \
---seed 124;
+--seed 124; 
+}} || exit_status="false"
```
Comment (ivadym) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_cnv_tumor_normal_tga.rule:208]: I checked more in depth the previous implementation and we were using the `exit_status` to set the default values of purity and ploidy. Here we are relying on the generated files, so I don't think defining this variable is necessary anymore

```suggestion
}} || true
```

**Comment by ivadym** [comment]
Has been replaced with https://github.com/Clinical-Genomics/BALSAMIC/pull/1406, closing!

**Pull request closed without merging**